text/0055-signature-based-finality.md

```diff
@@ -0,0 +1,115 @@
+
+[motivation]: #motivation
+
+The consensus component is currently quite heavy on memory, having to store the full protocol states of multiple eras from the past, in case some nodes decide to equivocate in those eras and have to be punished. Also, communicating the protocol state to other nodes is currently the only way to make them finalize blocks and advance to later eras if they are lagging behind. Thus, if we want to be able to drop older eras, we have to make sure that:
+
+- we can detect attempts to create a fork in the chain by equivocating, and
+- we can prove finality of the latest blocks to lagging nodes.
+
+Both of these goals can be achieved with signature based finality.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The current process of block finalization is this:
+
+1. A `BlockPayload` gets finalized by a consensus instance.
+2. The consensus component announces a `FinalizedBlock`.
+3. Contract runtime executes the block and announces a `Block`.
+4. Linear chain component stores the block and announces that it was stored.
+5. Consensus creates a finality signature for the block.
+6. Linear chain stores the signature and gossips it.
+
+In order to only consider fully signed blocks as finalized, we'll have to change this process to the following:
+
+1. A `BlockPayload` gets finalized by a consensus instance.
+2. The consensus component announces a `FinalizedBlock`.
+3. Contract runtime executes the block and announces a `Block`.
+4. In response to the announcement, linear chain component caches the block, and consensus creates a finality signature.
+5. Linear chain stores the signature and gossips it. It also receives signatures sent over the network.
+6. Once the linear chain gathers enough signatures for the block, it considers it finalized.
+    - If the block is unknown yet, it requests the block from the sender of the last signature.
+    - If the block is known, but it isn't the immediate descendant of the highest block so far, it requests the range of blocks from the immediate descendant of the highest block to the just finalized block.
+    - If the block is known and it is the immediate descendant of the highest block so far, it adds the block to the chain as the new highest block.
+
+A node receiving the request sent in point 6 would respond with the following:
+- To the request for a single fully signed block, send the block along with signatures proving it is fully signed.
+- To the request for a range of fully signed blocks, send the fully signed blocks one by one, starting from the oldest one.
+
+Upon receiving a block with signatures as a response:
+    - If it's not fully signed, discard and re-request it from someone else.
+    - If it's the immediate descendant of the highest known block:
+        - Request execution by the contract runtime.
+        - Compare the execution results to the block - if they don't match, discard the block and re-request it from someone else.
```

Comment:
If we're here, it means we received a block that is an immediate descendant of our last known/finalized/stored block, and the block we have just received has >2/3 of finality signatures. Should we really expect that other nodes will give us a block with the different post-execution results? Even if they did, this can't possibly sum up to the required 2/3 since we had just rejected a block signed by the quorum this big. If it does though, then it means there was an equivocation.

I think we should exit the node if the execution results of a block signed by more than 2/3 validators don't match.

cc @afck @xcthulhu

Reply:
That's a good point, I agree.